### PR TITLE
Fix .app population for single-unit peer relations

### PR DESCRIPTION
**Pull request by dshcherb**
Currently .app property is None when a peer relation contains only one
unit (relation-list returns an empty list). This change addresses the
issue.

### CODE REVIEW

File: ops/model.py [hunk 1]
```diff
@@ -236,6 +236,10 @@ def __init__(self, relation_name, relation_id, our_unit, backend, cache):
         except RelationNotFoundError:
             # If the relation is dead, just treat it as if it has no remote units.
             pass
+        # If an application has not been determined based on a remote unit, the only remaining
+        # case is a single unit and its peer relation where the application is our unit's application.
+        if self.app is None:
+            self.app = our_unit.app
```
Comment (jameinel) [inline ops/model.py]: On the Juju side, we've been talking about having a relation-created/relation-joined with the app tag, or some other indication to the unit that there *is* a relation, just no remote units have been started.
So while this is probably true in your peer relation case, I don't know if we can trust it to always be true. I realize that we don't yet do the work on the Juju side, but it seems fairly likely to end up in the 2.8 series. (we want to have a nice way to not get Blocked when you have a charm that is related, but the remote units haven't come up yet.)
Thoughts?

One option would be to look for JUJU_APPLICATION_NAME, another option would be to look for metadata relation.peer

Comment (johnsca) [reply ops/model.py]: I agree that this is likely to break for non-peer relations when we do get the `relation-created` hook.  However, I don't think we can use `JUJU_APPLICATION_NAME` or any other relation environment variable, because these lists will also be populated outside of relation hook contexts.

Some other options:

* Check the relation in question against the metadata to see if it's a peer relation (probably hard to reach the meta from this location without some refactoring)
* Get the application name during the relation hook via an env var and then persist it ourselves
* Depend on Juju to provide the application name consistently from something other than (or as an additional field on) `relation-list`

Comment (dshcherb) [reply ops/model.py]: I agree with @johnsca regarding `JUJU_APPLICATION_NAME` as I encountered this exactly in the case described when the event wasn't a relation event.

The use-case is a replacement of leader data when you put something into the peer app relation data bag in a non-relation event.

I also considered looking at a list of peer relations via charm metadata but, yes, this will require more significant changes.

It seems right to go down the route of inspecting charm metadata because we have access to it locally and it is deterministic. Quite possibly `is_peer` could be a variable passed into `__init__` of a `Relation` object as an alternative to passing metadata of the whole charm.

@jameinel @johnsca I could suggest a change with `is_peer` if you agree in general.

Comment (dshcherb) [reply ops/model.py]: Modified this based on relation roles: https://github.com/canonical/operator/pull/107/commits/da505df72b044a39f059c5444337b27b1f4f2be8

We could also change role names to "provider", "requirer", "peer" similar to how this is done in `juju/charm` https://github.com/juju/charm/blob/b0d0ce63cbe0c6e31d38e886a8c7cb20979b509d/meta.go#L42-L46

File: test/test_model.py [hunk 2]
```diff
@@ -361,7 +351,11 @@ def check_remote_units():
         # Create a new model and backend to drop a cached is-leader output.
         self.backend = ops.model.ModelBackend()
         meta = ops.charm.CharmMeta()
-        meta.relations = {'db0': None, 'db1': None, 'db2': None}
+        meta.relations = {
```
Comment (jameinel) [inline test/test_model.py]: Why was it ok before to have None (presumably it was populating it when it was needed, given the __init__ of Model also sets everything to None)?

File: ops/model.py [hunk 3]
```diff
@@ -42,7 +43,7 @@ def get_relation(self, relation_name, relation_id=None):
                     return relation
             else:
                 # The relation may be dead, but it is not forgotten.
-                return Relation(relation_name, relation_id, self.unit, self._backend, self._cache)
+                return Relation(relation_name, relation_id, self._meta.relations[relation_name].role, self.unit, self._backend, self._cache)
```
Comment (jameinel) [inline ops/model.py]: This does feel a bit strange. How is self.relations None, but we still have meta information about it?

This also *looks* like role is the odd-attribute-out. But it does feel like a property of a Relation to me.

File: ops/model.py [hunk 4]
```diff
@@ -217,16 +219,23 @@ def __getitem__(self, relation_name):
         if relation_list is None:
             relation_list = self._data[relation_name] = []
             for relation_id in self._backend.relation_ids(relation_name):
-                relation_list.append(Relation(relation_name, relation_id, self._our_unit, self._backend, self._cache))
+                relation_role = self._relations_meta[relation_name].role
```
Comment (jameinel) [inline ops/model.py]: You don't have to look this up in the loop, right? It doesn't change based on the relation-id.

File: ops/model.py [hunk 5]
```diff
@@ -217,16 +219,23 @@ def __getitem__(self, relation_name):
...
 class Relation:
-    def __init__(self, relation_name, relation_id, our_unit, backend, cache):
+    def __init__(self, relation_name, relation_id, role, our_unit, backend, cache):
         self.name = relation_name
         self.id = relation_id
+        self.role = role
         self.app = None
         self.units = set()
+
+        # For peer relations an app of a remote unit is a local app so the remote app becomes a local app. When there
```
Comment (jameinel) [inline ops/model.py]: # For peer relations, the both the remote and the local app are the same.

File: ops/model.py (hunk 3 above)
Comment (dshcherb) [reply ops/model.py]: `self.relations` may be present but the relation may be gone (i.e. no relation with such a relation id per `relation-ids`).

The weirdness here is also in the [for/else construct](https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops). 

`else` clause can be reached by using `get_relation` with a specific `relation_id` passed in while this relation id is gone at the time of a hook execution. So, a relation can be dead, but there will still be static metadata about it.

File: ops/model.py (hunk 4 above)
Comment (dshcherb) [reply ops/model.py]: Right, thanks for noticing.

File: test/test_model.py (hunk 2 above)
Comment (dshcherb) [reply test/test_model.py]: `RelationMeta` was not used in `model.py` for anything. There was no code that would access `self.relations['foo'].<attribute>`.

File: test/test_model.py [hunk 6]
```diff
@@ -81,6 +86,20 @@ def test_get_relation(self):
...
+        self.model = ops.model.Model('myapp/0', meta, self.backend)
+
+        err_msg = "ERROR invalid value \"$2\" for option -r: relation not found"
+        fake_script(self, 'relation-ids',
+                    """([ "$1" = dbpeer ] && echo '["dbpeer:0"]') || echo '[]'""")
+        fake_script(self, 'relation-list',
+                    f"""([ "$2" = 0 ] && echo '[]') || (echo {err_msg} >&2 ; exit 2)""")
+
+        db1_4 = self.model.get_relation('dbpeer')
+        self.assertEqual(db1_4.app, self.model.app)
```
Comment (johnsca) [inline test/test_model.py]: I don't know if it matters, but a potentially stronger assertion here would be `assertIs`

File: test/test_model.py [hunk 7]
```diff
@@ -118,11 +137,6 @@ def test_unit_relation_data(self):
         ])
 
     def test_remote_app_relation_data(self):
-        self.backend = ops.model.ModelBackend()
-        meta = ops.charm.CharmMeta()
-        meta.relations = {'db0': None, 'db1': None, 'db2': None}
-        self.model = ops.model.Model('myapp/0', meta, self.backend)
```
Comment (niemeyer) [inline test/test_model.py]: Why are these completely irrelevant, and copy & pasted so much? Error?

Comment (dshcherb) [reply test/test_model.py]: We used to have some tests that used fake tools with a real model backend instance and some that used [FakeModelBackend](https://github.com/canonical/operator/pull/61/files#diff-2bdcc3da19a2e638a126aab53c420862L20). Because of that, some test cases had to replace `self.backend` initialized to `FakeModelBackend` with an instance of `ModelBackend`. The point of #61 was to make all test cases use fake hook tools which it did but some test cases still had the `ModelBackend` initialization logic in-place which is no longer relevant.

In short: this particular change is here to clean up old test cases and move the common initialization code to `setUp`.

File: ops/model.py [hunk 8]
```diff
@@ -216,17 +218,23 @@ def __getitem__(self, relation_name):
...
+                relation_list.append(Relation(relation_name, relation_id, relation_role, self._our_unit, self._backend, self._cache))
         return relation_list
 
 
 class Relation:
-    def __init__(self, relation_name, relation_id, our_unit, backend, cache):
+    def __init__(self, relation_name, relation_id, role, our_unit, backend, cache):
         self.name = relation_name
         self.id = relation_id
+        self.role = role
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply ops/model.py]: I hope this covers it:
https://github.com/canonical/operator/pull/107/commits/3fa25dbc92551d0f0b159fbcd5467149fec1ece6

File: ops/model.py [hunk 9]
```diff
@@ -43,7 +43,8 @@ def get_relation(self, relation_name, relation_id=None):
                     return relation
             else:
                 # The relation may be dead, but it is not forgotten.
-                return Relation(relation_name, relation_id, self._meta.relations[relation_name].role, self.unit, self._backend, self._cache)
+                our_app = self.app if self._meta.relations[relation_name].role == 'peers' else None
```
Comment (niemeyer) [inline ops/model.py]: The straight version of this is much easier to read:
```
if self._meta.relations[relation_name].role == 'peers'
    our_app = self.app
else:
    our_app = None
```
Also, this logic seems to be distributed in two different places without benefit. Instead of doing it here, and then having a new _meta attribute, and then handing the underlying relations value down into relations which is also stored, let's move all of this into the relations value itself. We might call an underlying private method in the RelationMapping value, but instead of:
```
model.get_relation(name, id)
```
I think we can simply have:
```
model.relations.get(name, id)
```
With id still defaulting to None. Seems nice and follows a recognized pattern.

File: ops/model.py [hunk 10]
```diff
@@ -218,23 +219,22 @@ def __getitem__(self, relation_name):
         relation_list = self._data[relation_name]
         if relation_list is None:
             relation_list = self._data[relation_name] = []
-            relation_role = self._relations_meta[relation_name].role
+            our_app = self._our_unit.app if self._relations_meta[relation_name].role == 'peers' else None
```
Comment (niemeyer) [inline ops/model.py]: Related to the first comment. I suggest avoiding ternary operators in such cases.

File: ops/model.py [hunk 11]
```diff
@@ -218,23 +219,22 @@ def __getitem__(self, relation_name):
...
         self.id = relation_id
-        self.role = role
         self.app = None
         self.units = set()
 
         # For peer relations, both the remote and the local app are the same.
-        if self.role == 'peers':
-            self.app = our_unit.app
+        if our_app is not None:
+            self.app = our_app
```
Comment (niemeyer) [inline ops/model.py]: Instead of
```
self.app = None
(...)
if our_app is not None:
    self.app = our_app
```
This can be simply:
```
self.app = our_app
```
But! This shows a real problem in my recommendation: this is awkward because this is not our app. This is the relations app.. it makes no sense to tell the relation what the app is only sometimes.

Your suggestion of having an `is_peer` attribute seems more reasonable now. Sorry for taking you into this misleading path.

File: ops/model.py [hunk 12]
```diff
@@ -197,11 +199,12 @@ def __getitem__(self, key):
...
     """Map of relation names to lists of Relation instances."""
 
-    def __init__(self, relation_names, our_unit, backend, cache):
+    def __init__(self, relations_meta, our_unit, backend, cache):
+        self._relations_meta = relations_meta
         self._our_unit = our_unit
         self._backend = backend
         self._cache = cache
-        self._data = {relation_name: None for relation_name in relation_names}
+        self._data = {relation_name: None for relation_name in list(relations_meta)}
```
Comment (niemeyer) [inline ops/model.py]: Why transforming relations_meta into a list before iterating over it?

File: ops/model.py [hunk 13]
```diff
@@ -197,11 +199,12 @@ def __getitem__(self, key):
 class RelationMapping(Mapping):
     """Map of relation names to lists of Relation instances."""
 
-    def __init__(self, relation_names, our_unit, backend, cache):
+    def __init__(self, relations_meta, our_unit, backend, cache):
+        self._relations_meta = relations_meta
```
Comment (niemeyer) [inline ops/model.py]: Same here.. we're also changing from getting some data we just iterate over to build the type into storing part of the meta map. Yet, we don't actually use the map much. It'd be faster and nice to use if we did something like this I think:
```
self._peers = set()
for name, data in relations_meta:
    if data.role == 'peers':
        self._peers.add(name)
```
Now `name in self._peers` is all we need.

File: ops/model.py (hunk 11 above)
Comment (niemeyer) [reply ops/model.py]: Alternatively, we might have relation_app as a parameter, instead of our_app.

File: ops/model.py (hunk 12 above)
Comment (dshcherb) [reply ops/model.py]: Agreed, no reason to do this.

File: ops/model.py (hunk 13 above)
Comment (dshcherb) [reply ops/model.py]: Yes, this feels cleaner, thanks for spotting.

File: ops/model.py (hunk 9 above)
Comment (dshcherb) [reply ops/model.py]: The problem with `model.relations.get(name, id)` is that the default mixin doesn't expect the second `id` argument so simply extending `__getitem__` with the second argument doesn't work because the relevant parameter is simply not passed down by the following code:

https://github.com/python/cpython/blob/3.7/Lib/_collections_abc.py#L657-L662

One way to work around this would be making the key compound.

```
self.relations[(name, id)]
```

But then we lose the simplicity of `self.relations[name]`. Supporting both brings up the thought of supporting `self.relations[id]` as well but we currently do not have any facilities to lookup just by a relation id without specifying a relation name as this requires doing a reverse lookup over all relation names to retrieve a relation name to be placed into the `RelationMap`.

Comment (dshcherb) [reply ops/model.py]: Here is a rework of the relation object lookup logic:
https://github.com/canonical/operator/pull/107/commits/127e1f19a2424c99dfafdc3ac43f33576df7a567

Nothing better came to mind today than supporting compound keys in `get`, however, let's discuss this tomorrow and see if there's a better way.

Comment (dshcherb) [reply ops/model.py]: Got rid of compound keys per our discussion with @niemeyer and @chipaca. See:
https://github.com/canonical/operator/pull/107/commits/38bd46da91afe6b13c733d880cfee3147d8c9393

File: ops/model.py [hunk 14]
```diff
@@ -213,20 +199,48 @@ def __iter__(self):
...
-                relation_list.append(Relation(relation_name, relation_id, self._our_unit, self._backend, self._cache))
+            for rid in self._backend.relation_ids(relation_name):
+                relation = Relation(relation_name, rid, is_peer, self._our_unit, self._backend, self._cache)
+                relation_list.append(relation)
         return relation_list
 
+    def _get_unique(self, relation_name, relation_id=None):
+        if not isinstance(relation_id, (int, type(None))):
+            raise ModelError(f'relation name {relation_id} must be int or None not {type(relation_id).__name__}')
+        if relation_id is not None:
```
Comment (chipaca) [inline ops/model.py]: I'd suggest moving the first `if` inside the `None` if, as that simplifies it: that is,
```python
if relation_id is not None:
    if not isinstance(relation_id, int):
        raise ModelError('...')
    # etc
```

File: ops/model.py [hunk 15]
```diff
@@ -213,20 +199,48 @@ def __iter__(self):
...
+            for relation in self[relation_name]:
+                if relation.id == relation_id:
+                    return relation
+            else:
+                # The relation may be dead, but it is not forgotten.
+                is_peer = relation_name in self._peers
+                return Relation(relation_name, relation_id, is_peer, self._our_unit, self._backend, self._cache)
+        else:
```
Comment (chipaca) [inline ops/model.py]: this might be my recent time with Go showing: is the `else` necessary, here? I don't remember if that's pythonic or not :-D

Comment (dshcherb) [reply ops/model.py]: I suppose I could just drop it since every branch under the first if contains its own return statement or raises an exception.

Comment (dshcherb) [reply ops/model.py]: https://github.com/canonical/operator/pull/107/commits/6a058cb1cd04cb7f3d4ea38c01fd13a04d95cd03

File: ops/model.py (hunk 14 above)
Comment (dshcherb) [reply ops/model.py]: Thanks, that indeed simplifies it.

https://github.com/canonical/operator/pull/107/commits/6a058cb1cd04cb7f3d4ea38c01fd13a04d95cd03
